File: app/src/main/java/io/github/marcocipriani01/telescopetouch/activities/ImageDisplayActivity.java
/*
 * Copyright 2021 Marco Cipriani (@marcocipriani01) and the Sky Map Team
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.github.marcocipriani01.telescopetouch.activities;

import android.app.SearchManager;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.SharedPreferences.Editor;
import android.os.Bundle;
import android.util.Log;
import android.view.MenuItem;
import android.widget.Button;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.preference.PreferenceManager;

import io.github.marcocipriani01.livephotoview.PhotoView;
import io.github.marcocipriani01.telescopetouch.R;
import io.github.marcocipriani01.telescopetouch.TelescopeTouchApp;
import io.github.marcocipriani01.telescopetouch.activities.fragments.GoToFragment;
import io.github.marcocipriani01.telescopetouch.activities.util.DarkerModeManager;
import io.github.marcocipriani01.telescopetouch.gallery.GalleryImages;
import io.github.marcocipriani01.telescopetouch.layers.MessierLayer;
import io.github.marcocipriani01.telescopetouch.layers.PlanetsLayer;
import io.github.marcocipriani01.telescopetouch.layers.StarsLayer;

/**
 * Shows an image to the user and allows them to search for it.
 *
 * @author John Taylor
 */
public class ImageDisplayActivity extends AppCompatActivity {

    private static final String TAG = TelescopeTouchApp.getTag(ImageDisplayActivity.class);
    private static final int ERROR_MAGIC_NUMBER = -1;
    private GalleryImages selectedImage;
    private DarkerModeManager darkerModeManager;

    @Override
    protected void onCreate(Bundle icicle) {
        super.onCreate(icicle);
        ((TelescopeTouchApp) getApplication()).getApplicationComponent().inject(this);
        darkerModeManager = new DarkerModeManager(this, null, PreferenceManager.getDefaultSharedPreferences(this));
        setTheme(darkerModeManager.getPref() ? R.style.DarkerAppTheme : R.style.AppTheme);
        setContentView(R.layout.activity_gallery_image);
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
            actionBar.setDisplayShowHomeEnabled(true);
        }

        Intent intent = getIntent();
        Log.d(TAG, intent.toString());
        int position = intent.getIntExtra(ImageGalleryActivity.IMAGE_ID, ERROR_MAGIC_NUMBER);
        if (position == ERROR_MAGIC_NUMBER) {
            Log.e(TAG, "No position was provided with the intent - aborting.");
            finish();
        }

        selectedImage = GalleryImages.values()[position];
        PhotoView photo = this.findViewById(R.id.gallery_image);
        photo.setMaximumScale(5f);
        photo.setImageResource(selectedImage.getImageId());
        this.<TextView>findViewById(R.id.gallery_image_title).setText(selectedImage.getName(this));
        this.<Button>findViewById(R.id.gallery_image_search_btn).setOnClickListener(source -> {
            Log.d(TAG, "Do Search");
            // We must ensure that all the relevant layers are actually visible or the search might fail. This is rather hacky.
            SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(this);
            Editor editor = preferences.edit();
            String[] keys = {StarsLayer.PREFERENCE_ID, MessierLayer.PREFERENCE_ID, PlanetsLayer.PREFERENCE_ID};
            for (String key : keys) {
                if (!preferences.getBoolean(key, false)) {
                    editor.putBoolean(key, true);
                }
            }
            editor.apply();
            Intent queryIntent = new Intent(this, SkyMapActivity.class);
            queryIntent.setAction(Intent.ACTION_SEARCH);
            queryIntent.putExtra(SearchManager.QUERY, selectedImage.getSearchTerm(this));
            startActivity(queryIntent);
        });
        Button gotoButton = this.findViewById(R.id.gallery_point_telescope);
        String gotoName = selectedImage.getGotoName(this);
        gotoButton.setEnabled(!gotoName.equals("?"));
        gotoButton.setOnClickListener(source -> {
            Intent gotoIntent = new Intent(this, MainActivity.class);
            if (TelescopeTouchApp.connectionManager.isConnected()) {
                GoToFragment.setRequestedSearch(gotoName);
                gotoIntent.putExtra(MainActivity.ACTION, MainActivity.ACTION_SEARCH);
            } else {
                gotoIntent.putExtra(MainActivity.ACTION, MainActivity.ACTION_CONNECT);
                gotoIntent.putExtra(MainActivity.MESSAGE, R.string.connect_telescope_first);
            }
            gotoIntent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_NEW_TASK);
            startActivity(gotoIntent);
            finish();
        });
    }

    @Override
    public void onResume() {
        super.onResume();
        darkerModeManager.start();
    }

    @Override
    public void onPause() {
        super.onPause();
        darkerModeManager.stop();
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            this.finish();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }
}

File: app/src/main/java/com/google/android/stardroid/activities/ImageDisplayActivity.kt
// Copyright 2010 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
package com.google.android.stardroid.activities

import android.app.SearchManager
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.Button
import android.widget.ImageView
import android.widget.TextView
import androidx.preference.PreferenceManager
import com.google.android.stardroid.R
import com.google.android.stardroid.activities.DynamicStarMapActivity
import com.google.android.stardroid.activities.ImageDisplayActivity
import com.google.android.stardroid.activities.util.ActivityLightLevelChanger
import com.google.android.stardroid.activities.util.ActivityLightLevelManager
import com.google.android.stardroid.gallery.GalleryFactory
import com.google.android.stardroid.gallery.GalleryImage
import com.google.android.stardroid.util.Analytics
import com.google.android.stardroid.util.MiscUtil.getTag
import javax.inject.Inject

/**
 * Shows an image to the user and allows them to search for it.
 *
 * @author John Taylor
 */
class ImageDisplayActivity : InjectableActivity() {
    private var selectedImage: GalleryImage? = null
    private var activityLightLevelManager: ActivityLightLevelManager? = null

    @JvmField
    @Inject
    var analytics: Analytics? = null
    override fun onCreate(icicle: Bundle?) {
        super.onCreate(icicle)
        applicationComponent!!.inject(this)
        setContentView(R.layout.imagedisplay)
        activityLightLevelManager = ActivityLightLevelManager(
            ActivityLightLevelChanger(this, null),
            PreferenceManager.getDefaultSharedPreferences(this)
        )
        val intent = intent
        Log.d(TAG, intent.toString())
        val position = intent.getIntExtra(ImageGalleryActivity.IMAGE_ID, ERROR_MAGIC_NUMBER)
        if (position == ERROR_MAGIC_NUMBER) {
            Log.e(TAG, "No position was provided with the intent - aborting.")
            finish()
        }
        val galleryImages = GalleryFactory.getGallery(resources).galleryImages
        selectedImage = galleryImages[position]
        val imageView = findViewById<View>(R.id.gallery_image) as ImageView
        imageView.setImageResource(selectedImage!!.imageId)
        val label = findViewById<View>(R.id.gallery_image_title) as TextView
        label.text = selectedImage!!.name
        val backButton = findViewById<View>(R.id.gallery_image_back_btn) as Button
        backButton.setOnClickListener { source: View? -> goBack(source) }
        val searchButton = findViewById<View>(R.id.gallery_image_search_btn) as Button
        searchButton.setOnClickListener { source: View? -> doSearch(source) }
    }

    public override fun onStart() {
        super.onStart()
    }

    public override fun onResume() {
        super.onResume()
        activityLightLevelManager!!.onResume()
    }

    public override fun onPause() {
        super.onPause()
        activityLightLevelManager!!.onPause()
    }

    fun doSearch(source: View?) {
        Log.d(TAG, "Do Search")
        // We must ensure that all the relevant layers are actually visible or the search might
        // fail.  This is rather hacky.
        val sharedPreferences = PreferenceManager.getDefaultSharedPreferences(this)
        val editor = sharedPreferences.edit()
        val keys = arrayOf(
            "source_provider.0",  // Stars
            "source_provider.2",  // Messier
            "source_provider.3"
        ) // Planets
        for (key in keys) {
            if (!sharedPreferences.getBoolean(key, false)) {
                editor.putBoolean(key, true)
            }
        }
        editor.commit()
        val queryIntent = Intent()
        queryIntent.action = Intent.ACTION_SEARCH
        queryIntent.putExtra(SearchManager.QUERY, selectedImage!!.searchTerm)
        queryIntent.setClass(this@ImageDisplayActivity, DynamicStarMapActivity::class.java)
        startActivity(queryIntent)
    }

    fun goBack(source: View?) {
        Log.d(TAG, "Go back")
        finish()
    }

    companion object {
        private val TAG = getTag(ImageDisplayActivity::class.java)
        private const val ERROR_MAGIC_NUMBER = -1
    }
}